### ISSUE TITLE
InMemory database provider fails to execute query with enum discriminator

### ISSUE THREAD

**Issue by HKaresz** [dotnet/efcore#23926]
## Issue dscription

The unittest using the in-memory database provider fails to execute the query when the projection uses a class that has an enum discriminator. (Test1). In Test2, when the User's properties are not returned, the query succeeds.

When the discriminator is an int, I get a different error: `System.InvalidOperationException : The binary operator Equal is not defined for the types 'System.Nullable'1[System.Int32]' and 'System.Int32'.`

It also works well when the discriminator is string.

The SQL server provider executes the query properly.

The last EF package version, when the test succeeded was `5.0.0-preview.7.20365.15`.
After that (for `5.0.0-preview.8.20407.4` and `5.0.0-rc.1.20451.13`) I got `System.InvalidOperationException : The binary operator Equal is not defined for the types 'System.Nullable'1[EF_Inheritance_Issue.Entities.UserTypes]' and 'EF_Inheritance_Issue.Entities.UserTypes'.`

### Code

I've attached a cleaned-up sample application for repoducing the issue.
[https://1drv.ms/u/s!Ahz2yKeZswVtm_lmlyw-PaXWEYB0AA?e=NVBzhU](https://1drv.ms/u/s!Ahz2yKeZswVtm_lmlyw-PaXWEYB0AA?e=NVBzhU)

### Stack trace

Stack trace for the error:

```
  Message: 
    System.InvalidOperationException : Rewriting child expression from type 'EF_Inheritance_Issue.Entities.UserTypes' to type 'System.Nullable`1[EF_Inheritance_Issue.Entities.UserTypes]' is not allowed, because it would change the meaning of the operation. If this is intentional, override 'VisitUnary' and change it to allow this rewrite.
  Stack Trace: 
    ExpressionVisitor.ValidateChildType(Type before, Type after, String methodName)
    ExpressionVisitor.ValidateUnary(UnaryExpression before, UnaryExpression after)
    ExpressionVisitor.VisitUnary(UnaryExpression node)
    UnaryExpression.Accept(ExpressionVisitor visitor)
    ExpressionVisitor.Visit(Expression node)
    ExpressionVisitorUtils.VisitArguments(ExpressionVisitor visitor, IArgumentProvider nodes)
    ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
    NullableReadValueExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
    MethodCallExpression.Accept(ExpressionVisitor visitor)
    ExpressionVisitor.Visit(Expression node)
    NullableReadValueExpressionVisitor.VisitConditional(ConditionalExpression conditionalExpression)
    ConditionalExpression.Accept(ExpressionVisitor visitor)
    InMemoryQueryExpression.<AddLeftJoin>g__CopyEntityProjectionToOuter|38_2(EntityProjectionExpression entityProjection, Boolean nullable, <>c__DisplayClass38_0& )
    InMemoryQueryExpression.AddLeftJoin(InMemoryQueryExpression innerQueryExpression, LambdaExpression outerKeySelector, LambdaExpression innerKeySelector, Type transparentIdentifierType)
    InMemoryQueryableMethodTranslatingExpressionVisitor.TranslateLeftJoin(ShapedQueryExpression outer, ShapedQueryExpression inner, LambdaExpression outerKeySelector, LambdaExpression innerKeySelector, LambdaExpression resultSelector)
    QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
    InMemoryQueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
    MethodCallExpression.Accept(ExpressionVisitor visitor)
    ExpressionVisitor.Visit(Expression node)
    QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
    InMemoryQueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
    MethodCallExpression.Accept(ExpressionVisitor visitor)
    ExpressionVisitor.Visit(Expression node)
    QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
    InMemoryQueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
    MethodCallExpression.Accept(ExpressionVisitor visitor)
    ExpressionVisitor.Visit(Expression node)
    QueryableMethodTranslatingExpressionVisitor.TranslateSubquery(Expression expression)
    InMemoryExpressionTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
    MethodCallExpression.Accept(ExpressionVisitor visitor)
    ExpressionVisitor.Visit(Expression node)
    InMemoryExpressionTranslatingExpressionVisitor.TranslateInternal(Expression expression)
    InMemoryExpressionTranslatingExpressionVisitor.Translate(Expression expression)
    InMemoryProjectionBindingExpressionVisitor.Visit(Expression expression)
    InMemoryProjectionBindingExpressionVisitor.VisitNew(NewExpression newExpression)
    NewExpression.Accept(ExpressionVisitor visitor)
    ExpressionVisitor.Visit(Expression node)
    InMemoryProjectionBindingExpressionVisitor.Visit(Expression expression)
    InMemoryProjectionBindingExpressionVisitor.Translate(InMemoryQueryExpression queryExpression, Expression expression)
    InMemoryQueryableMethodTranslatingExpressionVisitor.TranslateSelect(ShapedQueryExpression source, LambdaExpression selector)
    QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
    InMemoryQueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
    MethodCallExpression.Accept(ExpressionVisitor visitor)
    ExpressionVisitor.Visit(Expression node)
    QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
    Database.CompileQuery[TResult](Expression query, Boolean async)
    QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)
    <>c__DisplayClass12_0`1.<ExecuteAsync>b__0()
    CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
    QueryCompiler.ExecuteAsync[TResult](Expression query, CancellationToken cancellationToken)
    EntityQueryProvider.ExecuteAsync[TResult](Expression expression, CancellationToken cancellationToken)
    EntityQueryable`1.GetAsyncEnumerator(CancellationToken cancellationToken)
    ConfiguredCancelableAsyncEnumerable`1.GetAsyncEnumerator()
    EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)
    UnitTest1.Test1() line 48
    --- End of stack trace from previous location ---

```

### Provider and version information

EF Core version: 5.0.2 (applies since 5.0.0-rc.2.20475.6)
Database provider: Microsoft.EntityFrameworkCore.InMemory
Target framework: NET 5.0
Operating system: Windows 10.18363 x64
IDE: Visual Studio 2019 16.8

**Comment by ajcvickers** [comment]
Note for triage: verified regression from 3.1. Still fails on latest daily, but not sure how up-to-date the daily builds are right now.

**Comment by Obi-Dann** [comment]
Encountered this issue as well while trying to upgrade to EF Core 5, most of the unit tests are failing because they use InMemory DB. Thanks for fixing it @smitpatel so promptly. Is there any workaround you can think of @smitpatel @ajcvickers while we're waiting for 5.0.4 release?

**Comment by smitpatel** [comment]
No work-around since it is deep inside the stack. 😞

**Comment by Obi-Dann** [comment]
Right, I found quite a dodgy way to get it working for our unit tests, but I definitely do NOT recommend using it in production.
There's a library called [Harmony](https://github.com/pardeike/Harmony) which often use to alter or mod functionality of video games
> Harmony gives you an elegant and high level way to alter the functionality in applications written in C#.

Sooo, writing a harmony patch like that gets the issue solved while waiting on the proper release.
Again, definitely don't recommend to use it in production!
```c#
[HarmonyPatch]
class NullableReadValueExpressionVisitorPatch
{
    /// <summary>
    /// Private internal EF class to patch
    /// </summary>
    static Type NullableReadValueExpressionVisitor =>
        typeof(InMemoryQueryExpression).Assembly.GetType("Microsoft.EntityFrameworkCore.InMemory.Query.Internal.InMemoryQueryExpression+NullableReadValueExpressionVisitor");

    /// <summary>
    /// Returns the method to patch
    /// </summary>
    static MethodInfo TargetMethod()
    {
        if (NullableReadValueExpressionVisitor.Assembly.GetName().Version >= Version.Parse("5.0.4.0"))
        {
            throw new Exception(@"Please delete this patch and remove references to Harmony if https://github.com/dotnet/efcore/issues/23926 is already fixed.
If not, then please update the version the issue is supposed to be fixed in");
        }

        return NullableReadValueExpressionVisitor.GetMethod("VisitUnary", BindingFlags.Instance | BindingFlags.NonPublic);
    }

    /// <summary>
    /// Runs before the target method.
    /// </summary>
    static bool Prefix(object __instance, UnaryExpression node, ref Expression __result)
    {
        if (__instance.GetType() == NullableReadValueExpressionVisitor)
        {
            // method body as per https://github.com/dotnet/efcore/pull/23984
            __result = node.Update((__instance as ExpressionVisitor).Visit(node.Operand));
            // returns false to skip calling the original method
            return false;
        }

        // returns true to call the original method
        return true;
    }
}
```

**Comment by mhamri** [comment]
+1 

I have this issue with OData 7.5+. confirmed with v8.0 preview also. 

I'm trying to update to efcore 5.0 but oData queries stopped working

**Comment by mhamri** [comment]
@ajcvickers this is my repo. I'm using sql instead of in-memory DB. 

https://github.com/mhamri/ef-int32-bug

**Comment by smitpatel** [comment]
@mhamri - Your issue is not same as this. Stacktrace is different. Please file a new issue.
